android/src/com/mapswithme/maps/location/LocationHelper.java

```diff
@@ -247,16 +248,19 @@ private void notifyLocationUpdated()
       mUiCallback.onLocationUpdated(mSavedLocation);
 
     // TODO: consider to create callback mechanism to transfer 'ROUTE_IS_FINISHED' event from
-    // the core to the platform code (https://jira.mail.ru/browse/MAPSME-3675),
+    // the core to the platform code (https://github.com/organicmaps/organicmaps/issues/3589),
     // because calling the native method 'nativeIsRouteFinished'
     // too often can result in poor UI performance.
     if (RoutingController.get().isNavigating() && Framework.nativeIsRouteFinished())
     {
       Logger.d(TAG, "End point is reached");
-      restart();
       if (mUiCallback != null)
         mUiCallback.onRoutingFinish();
       RoutingController.get().cancel();
+
+      // Break onLocationChanged() -> start() -> onLocationChanged() -> start() loop.
+      // https://github.com/organicmaps/organicmaps/issues/3588
+      runLater(this::restart);
```

Comment:
Is it really necessary to call restart here? Providers are working (if we are in notifyLocationUpdated), what should be changed here with restart call?

Reply:
I guess that this restart() is needed here to reduce GPS subscription interval after finishing navigation. I am not sure that it is essential.

Reply:
Looks like you are right. LGTM, anyway need refactor this with more natural way according to ToDo.